### PR TITLE
[2519] Add an mcb command for getting accredited courses

### PR DESCRIPTION
**Pull request by timabell**
### Context

We are working on understanding the needs of accredited bodies, in the short-term this will allow support to accurately and repeatably provide csv extracts to accredited bodies.

### Changes proposed in this pull request

* Extend provider model to give access to courses that a provider is the accredited body for.
* Add an mcb command
* Output is one row per course-site

### Not in this PR

* Easier to understand columns (currently exposes our raw model) - caused some confusion re draft / published etc.

### Guidance to review

Give it your worst!

```
mcb providers accredited_course_sites P80
mcb providers accredited_course_sites P80 -E qa
mcb providers accredited_course_sites P80 -f out.csv
```

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by timabell** [comment]
[Inaccurate data in ad hoc CSV's](https://trello.com/c/Mm9iX9Ap/2519-inaccurate-data-in-ad-hoc-csvs)

**Comment by timabell** [comment]
Code climate fail - quite correctly.

Going to extend test coverage for csv file output then refactor.

File: lib/mcb/commands/providers/accredited_course_sites.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  course_data = courses.select { |c| c[:sites].any? } # they aren't on "Find" or "Apply" if they have no sites
+  course_data = flatten_sites(course_data)
+  course_data.sort_by do |x|
+    [
+      x[:provider_name],
+      x[:course_name],
+      x[:program_type],
+      x[:qualification],
+      x[:site_name],
+    ]
```
Comment (misaka) [inline lib/mcb/commands/providers/accredited_course_sites.rb]: `values_at` could be tidier here.

File: lib/mcb/commands/providers/accredited_course_sites.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  course_data = get_course_data(provider)
+  if output_filename
+    write_to_csv_file(output_filename, course_data)
+  else
+    tp course_data
+  end
+end
+
+def get_course_data(provider)
+  courses = provider.current_accredited_courses.map do |c|
```
Comment (misaka) [inline lib/mcb/commands/providers/accredited_course_sites.rb]: How about this?

```
      if c.site_statuses.any?
        c.site_statuses.map do |ss|
          build_site(ss).merge(build_course(c))
        end
      else
        build_course(c)
      end
    end .sort_by { |c| c.values_at(...) }
  end

  def build_course(c)
    {
      provider_code: c.provider.provider_code,
      provider_name: c.provider.provider_name,
      course_code: c.course_code,
      course_name: c.name,
      study_mode: c.study_mode,
      program_type: c.program_type,
      qualification: c.qualification,
      content_status: c.content_status,
    }
  end

  def build_site(ss)
    {
      site_code: ss.site.code,
      site_name: ss.site.location_name,
      site_status: ss.status,
      site_published: ss.published_on_ucas?,
      site_vacancies: ss.vac_status,
    }
  end
```

File: lib/mcb/commands/providers/accredited_courses.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
+name "accredited_courses"
+summary "Show courses this provider is the accredited body for"
+param :code, transform: ->(code) { code.upcase }
+
+run do |opts, args, _cmd|
```
Comment (misaka) [inline lib/mcb/commands/providers/accredited_courses.rb]: How about we replace this with the `accredited_course_sites` command above?

File: spec/lib/mcb/commands/providers/accredited_course_sites_spec.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  let(:provider) { create :provider, :accredited_body }
+  let!(:findable_course) do
+    create :course, name: "findable-course",
+           accrediting_provider: provider,
+           site_statuses: [build(:site_status, :findable)]
+  end
+
+  it "writes course information to terminal" do
+    output = execute(arguments: [provider.provider_code])[:stdout]
+    expect(output).to include("provider_code | provider_name") # headers
```
Comment (misaka) [inline spec/lib/mcb/commands/providers/accredited_course_sites_spec.rb]: How about using `have_text_table_row`? Look for examples in existing code.

File: lib/mcb/commands/providers/accredited_courses.rb (hunk 3 above)
Comment (timabell) [reply lib/mcb/commands/providers/accredited_courses.rb]: done

File: spec/lib/mcb/commands/providers/accredited_course_sites_spec.rb (hunk 4 above)
Comment (timabell) [reply spec/lib/mcb/commands/providers/accredited_course_sites_spec.rb]: we can then validate all the column names without worrying about whitespace

File: lib/mcb/commands/providers/accredited_course_sites.rb (hunk 2 above)
Comment (timabell) [reply lib/mcb/commands/providers/accredited_course_sites.rb]: done

File: lib/mcb/commands/providers/accredited_course_sites.rb (hunk 1 above)
Comment (timabell) [reply lib/mcb/commands/providers/accredited_course_sites.rb]: done

**Pull request merged**